### PR TITLE
Fix ledger change to take into account cancelled message balance change

### PR DESCRIPTION
**Pull request by Leo-Besancon**
The ledger changes returned when settling the slot did not take into account the async message cancellation.

I'll try and add a TU to confirm everything is working

* [ ] document all added functions
* [ ] try in sandbox /simulation/labnet
  * [ ] if part of node-launch, checked using the `resync_check` flag
* [ ] unit tests on the added/changed features
  * [ ] make tests compile
  * [ ] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [ ] if the API has changed, update the API specification

### CODE REVIEW

File: massa-execution-worker/src/context.rs [hunk 1]
```diff
@@ -908,6 +908,13 @@ impl ExecutionContext {
             }
         }
 
+        // take the ledger changes again to take into account the balance change of canceled messages
+        let ledger_changes_canceled = self.speculative_ledger.take();
+
+        for (address, change) in ledger_changes_canceled.0 {
```
Comment (modship) [inline massa-execution-worker/src/context.rs:914]: haven't do rust for a while but this should works : 
```rust 
ledger_changes.0.extend(ledger_changes_canceled.0);
```

merge two collection

Comment (Leo-Besancon) [reply massa-execution-worker/src/context.rs:914]: It seems that we should use ledger_changes.apply(...) instead!

**Comment by bilboquet** [comment]
You may want to add a small UT to cover this case :)

**Comment by Leo-Besancon** [comment]
> You may want to add a small UT to cover this case :)

The problem here is that I didn't find a clean way to test the speculative ledger in the execution tests (the test universe seems to only have a handle on the final state). I'll dig a bit deeper on that

File: massa-execution-worker/src/context.rs (hunk 1 above)
Comment (Leo-Besancon) [reply massa-execution-worker/src/context.rs]: This is outdated, resolving as a result

File: massa-execution-worker/src/context.rs [hunk 2]
```diff
@@ -894,7 +894,7 @@ impl ExecutionContext {
         let deferred_credits_transfers = self.execute_deferred_credits(&slot);
 
         // take the ledger changes first as they are needed for async messages and cache
-        let ledger_changes = self.speculative_ledger.take();
+        let mut ledger_changes = self.speculative_ledger.get_snapshot();
```
Comment (bilboquet) [inline massa-execution-worker/src/context.rs:897]: is mut needed?

**Comment by bilboquet** [comment]
> The problem here is that I didn't find a clean way to test the speculative ledger in the execution tests (the test universe seems to only have a handle on the final state). I'll dig a bit deeper on that

If you can't find your way in UT, maybe it's doable with functional tests? (just an idea).

File: massa-execution-worker/src/context.rs [hunk 3]
```diff
@@ -894,7 +894,7 @@ impl ExecutionContext {
         let deferred_credits_transfers = self.execute_deferred_credits(&slot);
 
         // take the ledger changes first as they are needed for async messages and cache
```
[the remark this answers is not in the document]
Comment (damip) [reply massa-execution-worker/src/context.rs:896]: @Leo-Besancon I think it can be removed.

**Review by damip — APPROVED** [review]

**Comment by Leo-Besancon** [comment]
I'll prepare a functional test with a scenario to test both this PR and https://github.com/massalabs/massa/pull/4718.

**Pull request merged**